rust/ql/lib/codeql/rust/internal/Type.qll

```diff
@@ -15,10 +15,14 @@ newtype TType =
   TTrait(Trait t) or
   TArrayType() or // todo: add size?
   TRefType() or // todo: add mut?
+  TImplTraitType(int bounds) {
+    bounds = any(ImplTraitTypeRepr impl).getTypeBoundList().getNumberOfBounds()
+  } or
```

Comment:
If we have two functions like
```rust
fn f() -> impl MyTrait { ..  }

fn g() -> impl MyTrait { .. }
```
then they return two different types (in the general case) even though they are syntactically identical.

Given that wouldn't it be more accurate to represent `impl` types like:
```suggestion
  TImplTraitType(ImplTraitTypeRepr impl) or
```
I.e., every `impl` is it own types, similar to how every struct or enum is its own type. Then the `impl` types wouldn't need to have type parameters.

Reply:
Thanks for suggesting this, which I have implemented. I also learned that `impl Trait` types in parameter type positions are very different from `impl Trait` types in return type positions, so I have reworked the implementation to respect that (the former corresponds to a type parameter, the latter to an abstract type).